rse-r-package.Rmd

```diff
@@ -0,0 +1,629 @@
+
+FIXME: provide poorly-documented function and ask learner to find and fix gaps.
+
+### Exercise: Cross-references
+
+1.  What should you add to the roxygen2 comments for one function to link to the documentation for another function?
+2.  Add a cross-reference from the documentation for the function FIXME in the CO2 package to the documentation for the function FIXME.
+
+### Exercise: Documenting error conditions
+
+The guidelines above said that authors should document why and how their functions will deliberately fail.
+Where and how should you do this using roxygen2 for R?
+
+## How do I manage package dependencies? {#rse-r-package-dependencies}
+
+In order to understand the rest of what follows,
+it's important to understand that R packages are distributed as compiled [byte code](glossary.html#byte-code),
+*not* as [source code](glossary.html#source-code) (which is how Python does it)
+or as [binary code](glossary.html#binary-code) (which is how app stores distribute things).
+When a package is built,
+R loads and checks the code,
+then saves the corresponding low-level instructions.
+Our R files should therefore define functions,
+not run commands immediately:
+if they do the latter,
+those commands will be executed every time the script loads,
+which is probably not what users will want.
+
+As a side effect,
+this means that if a package uses `load(something)`,
+then that `load` command is executed *while the package is being compiled*,
+and *not* while the compiled package is being loaded by a user after distribution
+(Figure \@ref(fig:package-distribution)).
+If we have loaded the library by hand in our R session during development,
+though,
+we might not notice the problem.
+
+```{r package-distribution, echo=FALSE, fig.cap="Package Distribution"}
+knitr::include_graphics("figures/FIXME.png")
+```
+
+How then can our packages use other packages?
+The safest way is to use [fully-qualified names](glossary.html#fully-qualified-name)
+such as `stringr::str_replace`
+every time we call a function defined somewhere outside our package.
+Let's modify our word counter to use `stringr::str_count`:
+
+```{r word-count-stringr, eval=FALSE}
+word_count <- function(word, text) {
+  stringr::str_count(text, word)
+}
+```
+
+Since our compiled-and-distributable package will only contain the compiled code for its own functions,
+direct calls to functions from other packages won't work after the package is installed.
+To fix this,
+we ask `usethis` to add a note about `stringr` to `DESCRIPTION`:
+
+```{r use-package-stringr, eval=FALSE}
+usethis::use_package('stringr')
+```
+
+The bottom of `DESCRIPTION` now has these two lines:
+
+```
+Imports: 
+    stringr
+```
+
+The `Imports` field in `DESCRIPTION` tells R what else it has to install when it installs our package.
+Just to be on the safe side,
+we run `packageVersion("stringr")` in the console,
+which tells us that we have version 1.4.0 of `stringr`,
+and the modify `DESCRIPTION` to be:
```

Comment:
This is tedious when you have more than one package. Use `usethis::use_tidy_versions()` to add version numbers to all packages.